### ISSUE TITLE
Duplicate removals

### ISSUE THREAD

**Issue by ls-milkyway** [euzu/m3u-filter#139]
**Is your feature request related to a problem? Please describe.**
Different playlists may have duplicates channels in same or different groups ...how to filter out the duplicates based on "tvg-id" and/or "channel-name"...further problem becomes complicated as channel-names may be slightly different ...e.g "Reelz" and "Reelz by A&E" ...also the tvg-id can be different sometimes if playlist sources are from different countries and/or are using different EPG sources.

**Describe the solution you'd like**
Smartly remove duplicates based on preferred primary playlist or an option to select & keep the channel.

**Describe alternatives you've considered**
IPTVBoss has partial removal technique but only before importing any channels or groups.

**Additional context**
None.

**Comment by euzu** [comment]
@ls-milkyway there is no way to identify duplicate channels. Even when the namea are rqual it could be different channels.
The only way is to manually filter out duplicates. You can use a normal filter like
```
NOT (url ~ "^http...provider-to-exclude.com.*" AND (name ~ "dupl1" OR name ~ "dupl2"))
```

If you know a better way or an identification method for suplicates please tell me.

**Comment by ls-milkyway** [comment]
> @ls-milkyway there is no way to identify duplicate channels. Even when the names are equal it could be different channels. The only way is to manually filter out duplicates. You can use a normal filter like
> 
> ```
> NOT (url ~ "^http...provider-to-exclude.com.*" AND (name ~ "dupl1" OR name ~ "dupl2"))
> ```
> 
> If you know a better way or an identification method for suplicates please tell me.

@euzu Yep, no fool proof way ...like i said earlier ....assuming EPG providers are same for different sources of playlists, then "tvg-ids" will be same for both the playlists.

To conclude further "tvg-names" may be similar or slightly different e.g. "Reelz" and "Reelz by A&E" hence regular expression can be used on tvg-names to find the proximity.

**To make it foolproof:**
1) If streaming urls are same on 2 or more playlists or on the same playlist...then channels are 100% duplicates.
2) Check if EPG source url is same **and** "tvg-id" field on playlists are also same ...then channels are 100% duplicates (assuming epg & playlist provider are correct and precise).

ELSE

1) Check "tvg-names" are same or close using regular expression ....if yes then:
 2) Compare if "tvg-channel-logos" are same using AI image comparison.
 If all the 2 steps are true then channels I guess are 99% duplicates.

**Comment by euzu** [comment]
@ls-milkyway the problem with url is even the same channels can have different urls because of the provider. 
Only public channels which are listed on different public playlists can have the same url. 

I really dont have a starting point to realize this.   I could simply add some property comparisons, but i dont know if this will work as expected. And it must be appliable to for xtream and m3u playlists.

**Comment by ls-milkyway** [comment]
> @ls-milkyway the problem with url is even the same channels can have different urls because of the provider. Only public channels which are listed on different public playlists can have the same url.
> 
> I really dont have a starting point to realize this. I could simply add some property comparisons, but i dont know if this will work as expected. And it must be appliable to for xtream and m3u playlists.

@euzu
Yep, different providers .....different URLs ......but still some playlists generated or Concatenate using various sources like [https://github.com/iptv-org/](url) and [https://github.com/matthuisman/i.mjh.nz](url) may have duplicates from same provider plus some extra channels v...hence at least those will be filtered out to generated somewhat sanitized playlist.
The following may not be practically possible:

**Comment by euzu** [comment]
@ls-milkyway i have implemented a very simple remove duplicates  based on the `url`.  
To activate it you need to set the target option  `remove_duplicates` to `true`.


The image comparison using AI methods is too complex to be implemented in this project. The processing of the playlists must be completed within a reasonable timeframe.


Comparing names is also not effective. How should this work? First, compare the URL; if they don't match, then compare the names? However, the names for channels in different countries can be the same, for example: CNN or FOX can exist everywhere, but in different languages.

The use of EPG IDs is also not feasible. As you know, multiple sources can be merged into one. In this case, there are different EPG sources that may use the same IDs for different channels.


I currently have no idea how to further implement this. However, if you have any brilliant ideas, feel free to share them with me.

**Comment by ls-milkyway** [comment]
> @ls-milkyway i have implemented a very simple remove duplicates based on the `url`. To activate it you need to set the target option `remove_duplicates` to `true`.
> 
> The image comparison using AI methods is too complex to be implemented in this project. The processing of the playlists must be completed within a reasonable timeframe.
> 
> Comparing names is also not effective. How should this work? First, compare the URL; if they don't match, then compare the names? However, the names for channels in different countries can be the same, for example: CNN or FOX can exist everywhere, but in different languages.
> 
> The use of EPG IDs is also not feasible. As you know, multiple sources can be merged into one. In this case, there are different EPG sources that may use the same IDs for different channels.
> 
> I currently have no idea how to further implement this. However, if you have any brilliant ideas, feel free to share them with me.

Thanks for "remove duplicates"

Using AI cannot be practical....as stated earlier.

As stated earlier a url match is 100% duplicate. Further .....more deep search for removals .....since AI image comparison (is time consuming according to your info)...hence before executing it (if it was feasible) ....some criteria like comparing "tvg-names" and "tvg-ids" could be carried out...to conclude that AI image or AI image from broadcasting channels should be further compared or not.

Regarding "tvg-names" in many cases : for e.g."FOX" and "FOX-espanol" there is proximity in names (as stated earlier)....further in 99% cases user is going to select any one of them ...so things could have been carried out as mentioned in above paragraph and in case of 100% match user can be asked to select any 1 of it.

Nope, that's not the case multiple EPGs cannot have same IDs from different sources ..specially a good provider like [https://iptvboss.pro/](url)....like mac addresses.. the EPG providers are trying to to keep "tvg-id" as much unique as possible.

**Comment by euzu** [comment]
@ls-milkyway I will try to add functionality to compare some fields, which can be configured according to your needs.

**Comment by ls-milkyway** [comment]
> @ls-milkyway I will try to add functionality to compare some fields, which can be configured according to your needs.

Thanks....I would further like to state that this can be general need for a TV fanatic or a fan who like watching TV a lot with properly arranged channels. For e.g.

You can download m3u files from [https://github.com/iptv-org/iptv](url) for English speaking regions like US, CA, AU, NZ ....and process the final list by removing duplicates ....you may argue that it already has english language list....true, but firstly it's not complied properly secondly we have some other GITHUB projects with such country based list having same and better channels from better providers.

Hence it becomes easier for user to remove duplicates and pass it to IPTVboss etc. to form groups and arrange channels for better viewing experience.
Regards.
